scipy/integrate/_ivp/radau.py

```diff
@@ -105,8 +110,12 @@ def solve_collocation_system(fun, t, y, h, Z0, scale, tol,
         if not np.all(np.isfinite(F)):
             break
 
-        f_real = F.T.dot(TI_REAL) - M_real * W[0]
-        f_complex = F.T.dot(TI_COMPLEX) - M_complex * (W[1] + 1j * W[2])
+        if mass_matrix is None:
+            f_real = F.T.dot(TI_REAL) - M_real * W[0]
+            f_complex = F.T.dot(TI_COMPLEX) - M_complex * (W[1] + 1j * W[2])
+        else:
+            f_real = F.T.dot(TI_REAL) - M_real * mass_matrix.dot(W[0])
```

Comment:
Might want to make sure `mass_matrix` is an array or sparse array first. I think this will fail with array_like.

Reply:
Indeed, the Jacobian of the ODE/DAE system is transformed to a Numpy array if it is an array_like, via the `_validate_jac` function. I have added a similar `_validate_mass_matrix` function to handle all the required checks and transformations.